quixstreams/core/stream/stream.py

```diff
@@ -302,91 +299,101 @@ def diff(self, other: Self) -> Self:
-        for node in diff:
-            # Copy the node to ensure we don't alter the previously created Nodes
-            node = copy.deepcopy(node)
-            node.parent = parent
-            parent = node
-        self._prune(diff[0])
-
-        if parent is None:
-            raise InvalidOperation("No common parent found")
-
-        return parent
-
-    def root_path(self, allow_splits=True) -> List[Self]:
-        """
-        Return a list of all parent Streams including the node itself.
-
-        Can optionally stop at a first encountered split with allow_splits=False
-
-        The tree is ordered from parent to child (current node comes last).
-        :return: a list of `Stream` objects
-        """
-
-        node = self
-        tree_ = [node]
-        while (parent := node.parent) and (allow_splits or len(parent.children) < 2):
-            tree_.append(parent)
-            node = node.parent
-
-        # Reverse to get expected ordering.
-        tree_.reverse()
-
-        return tree_
+        # Cut off the diffed nodes from the rest of the tree
+        # by removing the diff head node the current node's children
+        diff_head.pruned = True
+        self.children.remove(diff_head)
+        diff_head.parents = []
+
+        # Copy the diff head and all nodes related to it
+        # to ensure the diff is fully extracted from the tree
+        # and not connected to other nodes.
+        diff_head = copy.deepcopy(diff_head)
+        parents = [diff_head]
+        for stream in diff[1:]:
+            stream = copy.deepcopy(stream)
+            stream.parents = parents
+            stream.children = copy.deepcopy(stream.children)
+            parents = [stream]
+        return diff_head
 
     def full_tree(self) -> List[Self]:
         """
-        Starts at tree root and finds every Stream in the tree (including splits).
-        :return: The collection of all Streams interconnected to this one
+        Find every related Stream in the tree across all children and parents and return
+        them in a topologically sorted order.
         """
-        return self._collect_nodes([], self.root_path()[0])
+        # Sort the tree in the topological order using graphlib.TopologicalSorter
+        # Topological sorting resolves all dependencies in the graph, and child nodes
+        # always come after all the parents
+        sorter: TopologicalSorter = TopologicalSorter()
+        visited: set[Self] = set()
+
+        # Add all the nodes to the TopologicalSorter
+        to_traverse: Deque[Self] = collections.deque()
+        to_traverse.append(self)
+        while to_traverse:
+            node = to_traverse.popleft()
+            if node in visited:
+                continue
+            sorter.add(node, *node.parents)
+            visited.add(node)
+            for parent in node.parents:
+                to_traverse.append(parent)
+            for child in node.children:
+                to_traverse.append(child)
```

Comment:
Consider something more concise
```suggestion
            to_traverse.extend(node.parents)
            to_traverse.extend(node.children)
```
or
```
to_traverse += node.parents + node.children
```